streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.processor.api.RecordMetadata;
+import org.apache.kafka.streams.query.Position;
+import org.apache.kafka.streams.query.PositionBound;
+import org.apache.kafka.streams.query.Query;
+import org.apache.kafka.streams.query.QueryResult;
+
+public final class StoreQueryUtils {
+
+    // make this class uninstantiable
+    private StoreQueryUtils() {
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <R> QueryResult<R> handleBasicQueries(
+        final Query<R> query,
+        final PositionBound positionBound,
+        final boolean collectExecutionInfo,
+        final StateStore store) {
+
+        final QueryResult<R> result;
+        final long start = collectExecutionInfo ? System.nanoTime() : -1L;
+        // TODO: position tracking
+        if (query instanceof PingQuery) {
+            result = (QueryResult<R>) QueryResult.forResult(true);
```

Comment:
How do we ensure type R matches the type of template type of the store?

Reply:
Unfortunately, it's not possible to guarantee that using Java's type system. The query caller will just have to be sure to submit correctly typed queries based on their a priori knowledge of the store's types. If they get it wrong, they'll get a ClassCastException at run time. This is a bummer, but it's also precisely the same in IQ today.

As far as I know, it's not possible to do better than this.